### PR TITLE
Saaly/enable disable devices

### PR DESCRIPTION
**Pull request by Sarah-Aly**
This should resolve [milestone no. 4](https://github.com/brentstineman/Transit-IOTHub-Demo/milestone/4)

### Work items
* Device Property/Action: Enable/Disable device #10
* Management API: Enable/Disable Device #9

### CODE REVIEW

File: src/Devices/Base/BaseDevice.cs [hunk 1]
```diff
@@ -64,5 +68,18 @@ public void SendMessageToCloud (string messageString)
 
         }
 
+        public async Task EnableDevice()
```
Comment (brentstineman) [inline src/Devices/Base/BaseDevice.cs]: Should this be a "setDeviceState" type of property that can be set to enabled or dissabled?

File: src/Functions/TransitFunctionApp/SetDeviceStatus.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "SetDeviceStatus/{DeviceId:guid}/{status}")] HttpRequest req,
+            string deviceId,
+            string status,
+            ILogger log)
+        {
+            DeviceStatus newStatus;
+            if (!Enum.TryParse(status, out newStatus)) {
+                return new BadRequestResult();
+            }
+            DeviceClient client = DeviceClient.CreateFromConnectionString(Environment.GetEnvironmentVariable("IotHubConnectionString"), deviceId);
```
Comment (brentstineman) [inline src/Functions/TransitFunctionApp/SetDeviceStatus.cs]: Do we want to validate the device ID? or are we confident that a simple "not found" will be sufficient?

**Review by brentstineman — COMMENTED** [review]
Can we add some unit tests for the new functionality? With that in place and the 2 comments addressed, we should be good to go. :)

File: src/Devices/Base/BaseDevice.cs (hunk 1 above)
Comment (Sarah-Aly) [reply src/Devices/Base/BaseDevice.cs]: Done!

File: src/Functions/TransitFunctionApp/SetDeviceStatus.cs (hunk 2 above)
Comment (Sarah-Aly) [reply src/Functions/TransitFunctionApp/SetDeviceStatus.cs]: I tried to check the online docs but couldn't find the contract for CreateFromConnectionString; so just wanted to be in the safe side & checked for null

Comment (Sarah-Aly) [reply src/Functions/TransitFunctionApp/SetDeviceStatus.cs]: I tried to check the online docs but couldn't find the contract for CreateFromConnectionString; so just wanted to be in the safe side & checked for null

File: src/Devices/Base/BaseDevice.cs [hunk 3]
```diff
@@ -68,18 +68,19 @@ public void SendMessageToCloud (string messageString)
 
         }
 
-        public async Task EnableDevice()
+        public async Task SetDeviceStatus(DeviceStatus status)
```
Comment (brentstineman) [inline src/Devices/Base/BaseDevice.cs:91]: Shouldn't this be SetDeviceStatusAsync?

**Comment by brentstineman** [comment]
ok @Sarah-Aly I looked over the code and sorry to keep bringing up more items. In addition to the comments above, I think we need 2 more changes. 

1) we need a unit test for the set status API. 
2) The devices initial state should be set based on its config. This also means updating the IDeviceConfig to include it as a default property. This also creates an interesting solution where when the device restarts, should it pull from the device twin, or go with the local config and just update its twin property. I've got a bit of this already set up in my PR here: https://github.com/brentstineman/Transit-IOTHub-Demo/blob/ChangeGateDirection/src/Devices/GateReader/GateReaderDevice.cs#L50-L62

I'll leave it to you to decide which should be the source of truth on a restart, the twin or the config. It can even be another config property, or a 3rd status. I trust whatever you decide to do.

File: src/Devices/Base/BaseDevice.cs (hunk 3 above)
Comment (Sarah-Aly) [reply src/Devices/Base/BaseDevice.cs:91]: Done!

File: src/Functions/TransitFunctionApp/SetDeviceStatus.cs (hunk 2 above)
Comment (Sarah-Aly) [reply src/Functions/TransitFunctionApp/SetDeviceStatus.cs]: @brentstineman , does this work? :)

**Comment by Sarah-Aly** [comment]
@brentstineman, I followed your example with 'gate direction'

File: src/Functions/TransitFunctionApp/SetDeviceStatus.cs (hunk 2 above)
Comment (brentstineman) [reply src/Functions/TransitFunctionApp/SetDeviceStatus.cs]: looks good!

**Pull request merged**